Add a PSR-3 logger and support any PSR-3 logger

pbowyer opened the pull request:
# Context and history

- In 2015 I [opened an RFC](https://github.com/modxcms/revolution/issues/12262) to explore adding PSR-3 logging support to MODX.
- This was superseded in 2016 by [MAB-05](https://github.com/modxcms/mab-recommendations/blob/master/05-psr3-logging-standard.md) which recommended [adopting the PSR-3 Logging Standard](https://github.com/modxcms/mab-recommendations/blob/master/05-psr3-logging-standard.md).
- In 2021 @Mark-H opened #213. While that branch remains open, it provided the conceptual foundation for this implementation.
- In late 2025 while developing custom MODX logging output to help with debugging, I realised there was no way to extend logging without modifying core files.

This PR builds on previous efforts to finally resolve those constraints while ensuring 100% backward compatibility.

# Goals
1. 100% backwards compatibility. xPDO users should not know anything has changed
2. Include a lightweight PSR-3 logger within xPDO, which preserves existing behavior
3. Enable any PSR-3 implementation to be registered with the service container

# Implementation
I have integrated psr/log with a simple default implementation. The logic follows a "opt-in" flow:

- If a custom PSR logger is provided, xPDO switches to PSR-3 output.
- Otherwise, it continues using the legacy xPDO logging logic.

I’ve included exhaustive tests in xPDOLoggingHistoricTest.php to ensure no regressions from the current behaviour.

I started by looking at @Mark-H's branch but I wanted to preserve full backwards compatibility, so used it as inspiration rather than building directly on it. In my code MODX logging works exactly as it currently does, until a custom PSR logger is provided. Once a custom PSR logger is provided, it's assumed that you want PSR-3 output and  you opt out of the traditional MODX logging.

# Request for feedback
- I have added a public property `xPDO::$logger`. Is this acceptable for ease of use, or should we strictly require the service container for access?
- Monolog is a dev dependency purely so the tests can use it as a 3rd-party implementation
- I added more code to `xPDO` than I wished to, and there is some duplication to support the current xPDO logging behaviours. Is this level of duplication acceptable to keep the "legacy" side untouched?
- Currently `xPDOLogger` has a dependency on `xPDO`. It works, but it feels unclean. Would you prefer a different approach?
- I opted not to mark old methods as @deprecated to keep this PR strictly about adding new features. Would you prefer I add those now?

# Remaining Todos
- [ ] Any duplicated tests in `xPDOLoggingHistoricTest.php` and `xPDOLoggerTest.php` need rationalising. I am satisfied that `xPDOLoggingHistoricTest.php` is exhaustive

pbowyer committed:
First pass

pbowyer committed:
Always initialise xPDOLogger; extend tests

pbowyer committed:
Revisions to clean xPDO

pbowyer committed:
Change composer.json to support PHP 7.2.5+

pbowyer committed:
Add logger to the container

pbowyer committed:
Test legacy logging

pbowyer committed:
Simplify logger initialization

`initConfig()` sorts out the data, so we don't have to look at `$options` when setting up the logger

opengeek commented:
Thanks for this work, Peter! I will check this out as soon as possible and get some feedback to you.

pbowyer committed:
Extend historic logging

pbowyer committed:
Fix regression in fatal logging, where the output was incorrect

pbowyer committed:
Re-enable xPDOLoggerTest

opengeek commented on src/xPDO/Logging/xPDOLogger.php:
```diff
@@ -0,0 +1,263 @@
+                $backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);
+            } else {
+                $backtrace = debug_backtrace();
+            }
```
We can remove this conditional logic as 5.4 is well below our support matrix.

opengeek commented on src/xPDO/xPDO.php:
```diff
@@ -2124,6 +2237,93 @@ protected function _log($level, $msg, $target= '', $def= '', $file= '', $line= '
+                $backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS);
+            } else {
+                $backtrace = debug_backtrace();
+            }
```
Again, I don't think we need this.

opengeek reviewed (commented):
I'm still considering the necessity of the new methods on the xPDO class. Are these potential overkill? I am wondering if we could simply activate it if a PSR-3 logger was provided in the container… 🤔

pbowyer committed:
Strip back additions to xPDO, use PSR3 logger only when in container

pbowyer committed:
Compact into one method

pbowyer commented:
I agree, how about something like this: https://github.com/modxcms/xpdo/compare/3.x...pbowyer:xpdo:3.x-psr-logger-take2

Or were you thinking more pared back?

opengeek commented:
> I agree, how about something like this: [3.x...pbowyer:xpdo:3.x-psr-logger-take2](https://github.com/modxcms/xpdo/compare/3.x...pbowyer:xpdo:3.x-psr-logger-take2)
> 
> Or were you thinking more pared back?

Yeah, that is looking more like what I had in mind.

pbowyer committed:
xPDO: remove unused resolveLogLocation helper

resolveLogLocation() in src/xPDO/xPDO.php was added but never used.

Keeping it around increases maintenance surface and risk: its logic overwrote any backtrace-derived file with $_SERVER['SCRIPT_NAME'], which would produce misleading log locations if it were ever called.

Remove the dead helper to avoid confusion and prevent accidental future use of incorrect location resolution.

pbowyer committed:
xPDO: reduce backtrace overhead when resolving log location

When callers omit $file/$line, xPDO::_log() resolves the log location using debug_backtrace().

The default debug_backtrace() captures the full stack (including args), which is unnecessary overhead when we only need the immediate caller's file/line.

Use debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 3) to keep the same frame indexing used by the existing logic while avoiding deep/arg-heavy traces. This preserves the resulting file/line values in log output while reducing CPU/memory cost.

pbowyer committed:
Merge branch '3.x-psr-logger-take2' into 3.x-psr-logger

pbowyer commented:
Thanks @opengeek, I've cleaned it up further and merged it into this branch

pbowyer committed:
Reduce backtrace overhead further

Extend the change in c6ab4d91746b69e178c7c97166e69eccb71f201a to xPDOLogger

pbowyer committed:
Allow psr/log ^3.0

pbowyer commented:
I've also done another "What if..." branch: https://github.com/modxcms/xpdo/compare/3.x...pbowyer:xpdo:3.x-psr-logger-take3

This lets the `xPDOLogger` class handle all legacy behaviour, streamlining `xPDO`.

I've included an optional BC break in a separate commit which removes `protected \xPDO\xPDO::_getLogLevel($level)`. I suspect other code not least MODX relies on this method, but it was too tempting to leave out.

Mark-H commented:
Lovely to see this moving!

JoshuaLuckers commented:
Nice work @pbowyer !

opengeek commented:
FYI @pbowyer — I am experimenting with Copilot reviews on your PR. I'm mostly curious to see the results. Do not take this as feedback to act on immediately.

pbowyer commented on src/xPDO/xPDO.php:
```diff
@@ -2047,6 +2066,61 @@ protected function _log($level, $msg, $target= '', $def= '', $file= '', $line= '
+            $backtrace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 2);
+            if ($backtrace && isset($backtrace[1])) {
+                $file = $backtrace[1]['file'];
+                $line = $backtrace[1]['line'];
```
@copilot code review[agent]  Now this one's interesting because the tests pass, and I wrote the tests against the parent branch. So did I get the tests wrong - or have they been silently modified?

opengeek commented:
Well, that was underwhelming in terms of feedback. 🤔

pbowyer commented:
> FYI @pbowyer — I am experimenting with Copilot reviews on your PR. I'm mostly curious to see the results. Do not take this as feedback to act on immediately.

No problem! Unlike the Gemini code review bot, it looks like with Copilot you can't have a conversation about the changes?

opengeek commented:
> > FYI @pbowyer — I am experimenting with Copilot reviews on your PR. I'm mostly curious to see the results. Do not take this as feedback to act on immediately.
> 
> No problem! Unlike the Gemini code review bot, it looks like with Copilot you can't have a conversation about the changes?

I thought you could by just mentioning it… hmm

pbowyer commented:
> > > FYI @pbowyer — I am experimenting with Copilot reviews on your PR. I'm mostly curious to see the results. Do not take this as feedback to act on immediately.
> > 
> > 
> > No problem! Unlike the Gemini code review bot, it looks like with Copilot you can't have a conversation about the changes?
> 
> I thought you could by just mentioning it… hmm

@opengeek can you try as (I assume) the bot's owner? [My mention](https://github.com/modxcms/xpdo/pull/265#discussion_r2860738376) broke into text...

opengeek commented:
> > > > FYI @pbowyer — I am experimenting with Copilot reviews on your PR. I'm mostly curious to see the results. Do not take this as feedback to act on immediately.
> > > 
> > > 
> > > No problem! Unlike the Gemini code review bot, it looks like with Copilot you can't have a conversation about the changes?
> > 
> > 
> > I thought you could by just mentioning it… hmm
> 
> @opengeek can you try as (I assume) the bot's owner? [My mention](https://github.com/modxcms/xpdo/pull/265#discussion_r2860738376) broke into text...

I guess I am wrong!

> Any comments you add to Copilot's review comments will be visible to humans, but they won't be visible to Copilot, and Copilot won't reply.

pbowyer commented:
cOpiLOt sUcKs

Another open source project I contribute to uses https://developers.google.com/gemini-code-assist/docs/review-github-code which has been rather helpful. Free up to 33 PR a day I believe. And OpenAI has their own too I believe.

opengeek commented:
> cOpiLOt sUcKs
> 
> Another open source project I contribute to uses https://developers.google.com/gemini-code-assist/docs/review-github-code which has been rather helpful. Free up to 33 PR a day I believe. And OpenAI has their own too I believe.

I'm going to reserve judgment for now. I had it review my PR yesterday and it was super insightful. I think you just wrote really good code and it is potentially confused on the backtrace level? 😅

pbowyer commented:
Mine has already been reviewed by AI multiple times before submitting (and I've got AI reviewing the backtrace level locally to confirm) so that could be the case!

Here's an example Gemini code assist review: https://github.com/BadgerHobbs/BinDays-API/pull/182#pullrequestreview-3862505759

opengeek commented:
> Mine has already been reviewed by AI multiple times before submitting (and I've got AI reviewing the backtrace level locally to confirm) so that could be the case!
> 
> Here's an example Gemini code assist review: [BadgerHobbs/BinDays-API#182 (review)](https://github.com/BadgerHobbs/BinDays-API/pull/182#pullrequestreview-3862505759)

Nice! This is the kind of AI-assisted contribution I love to see. 🙌🏼

pbowyer committed:
Bugfix: debug_backtrace() was off by 1

pbowyer commented:
Copilot was right about `debug_backtrace()` not being the same as before. I like to think I'd changed it for good reason, but I can't remember if I did 😊 so I've made it match the `3.x` behaviour. 

I've also updated the alternative 'take3' branch: https://github.com/modxcms/xpdo/compare/3.x...pbowyer:xpdo:3.x-psr-logger-take3 (comparison with this PR: https://github.com/pbowyer/xpdo/compare/3.x-psr-logger...pbowyer:xpdo:3.x-psr-logger-take3)

opengeek commented:
@pbowyer — I just realized we don't have a record of you signing the MODX CLA. Can you go to https://account.modx.com/, navigate to the Review CLA tab, and sign it? I'd like to get this integrated!

opengeek commented:
@pbowyer bump…

pbowyer commented:
Hi @opengeek back from holiday now... I signed the CLA in 2011 and I've forwarded the PDF to you via email. There's no clause in https://account.modx.com/cla that revokes a previously signed CLA so we'd need to revoke the 2011 one before I sign online, otherwise we've two CLAs in play!

opengeek committed:
Add docblocks to public methods
